google-cloud-spanner/src/main/java/com/google/cloud/spanner/AsyncResultSetImpl.java

```diff
@@ -458,16 +479,24 @@ public ApiFuture<Void> setCallback(Executor exec, ReadyCallback cb) {
           this.state == State.INITIALIZED, "callback may not be set multiple times");
 
       // Start to fetch data and buffer these.
-      this.result =
-          new ListenableFutureToApiFuture<>(this.service.submit(new ProduceRowsCallable()));
+      this.result = SettableApiFuture.create();
+      this.state = State.IN_PROGRESS;
+      this.service.execute(new InitiateStreamingRunnable());
       this.executor = MoreExecutors.newSequentialExecutor(Preconditions.checkNotNull(exec));
       this.callback = Preconditions.checkNotNull(cb);
-      this.state = State.RUNNING;
       pausedLatch.countDown();
       return result;
     }
   }
 
+  private void initiateProduceRows() {
+    this.service.execute(new ProduceRowsRunnable());
+    if (this.state == State.IN_PROGRESS) {
```

Comment:
This is to avoid to overriding of state. For example, resultset.cancel can be called before `initiateProduceRows` called. In such cases, this will not override.

Reply:
We don't need it as `initiateProduceRows` is being called from inside of `synchronized (monitor)` in `onStreamMessage`